### PR TITLE
Inventory reader

### PR DESCRIPTION
**Pull request by manpreet85**
The code had more than 100 problems. Please help

### CODE REVIEW

File: internal/config/InventoryReader_test.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	require.Equal(t, 2, len(response.Racks))
+
+// Racks is a map of a maps (A struct or array type). for eg: Rack1 is an map of blade 1 and blade2. blade 1 itself is a map.
+	
+	r,ok:= response.Racks["rack1"] // reading array value of  map Racks that was captured in response into a new variable r. 
+	require.True(t, ok) // Continue only if the boolean value is True. i.e rack1 was found.
+	assert.Equal(t, 2, len(r.Blades)) //Confirming if the length of array item r.Blades is 2. i.e there are two blades in rack1.
+	
+	b,ok:=r.Blades[1] //capturing key blade 1 into the variable b. The second value is a boolean which indicates if the key was successfully found  or not.
+	require.True(t, ok)//continue only if the boolean value is True. i.e Blade 1 was found.
```
Comment (Jim3Things) [inline internal/config/InventoryReader_test.go:48]: I don't think you need these comments, but if you do, then they should be on their own lines.

The first comment seems off, but I'm not sure.  At this point there is a map of known racks in response.Racks.  We look up 'rack1' in line 43, and verify that it exists in the Racks collection in line 44.

File: internal/config/InventoryReader_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+	unit_test.SetTesting(t)
+	defer unit_test.SetTesting(nil)
+
+	response, err := ReadInventoryDefinition(".")
+	require.Nil(t, err)
+
+	require.Equal(t, 2, len(response.Racks))
+
+// Racks is a map of a maps (A struct or array type). for eg: Rack1 is an map of blade 1 and blade2. blade 1 itself is a map.
```
Comment (Jim3Things) [inline internal/config/InventoryReader_test.go:41]: I'm not sure that I follow what this is saying.  Racks is a map of Rack objects, Each Rack contains a set of attributes, one of which is a map of Blades held by that rack.  Each blade is an object that contains a set of attributes (and no map).

File: internal/config/InventoryReader_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	assert.Equal(t, int64(2048), c.NetworkBandwidthInMbps)
+}
+
+func TestReadInventoryBogusPath (t *testing.T){
+	unit_test.SetTesting(t)
+	defer unit_test.SetTesting(nil)
+
+	response, err := ReadInventoryDefinition("C://Users//Waheguru") //Points to the user directory that do not have the YAML File
+	require.NotNil(t, err) //Checks that the error is not empty. IF it is than it should throw back an error
+	assert.NotEqual(t, "%v", response)//This is just so that I can use variable response. No actual impact on the test.
```
Comment (Jim3Things) [inline internal/config/InventoryReader_test.go:73]: Make line 71 be '_, err := '.  You don't care about response here, and can thus eliminate line 73.  I'd also really suggest not using your personal directory path in the code.  Pick something that can't work, say, "..".  If you can come up with something that must be imaginary on all systems, that would work too.

File: internal/services/frontend/inventory.go [hunk 4]
```diff
@@ -185,3 +185,4 @@ func handlerBladeRead(w http.ResponseWriter, r *http.Request) {
 		return p.Marshal(w, blade)
 	})
 }
+
```
Comment (Jim3Things) [inline internal/services/frontend/inventory.go:188]: Revert the changes in this file to remove it from this checkin.

File: internal/config/InventoryReader_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	defer unit_test.SetTesting(nil)
+
+	response, err := ReadInventoryDefinition(".")
+	require.Nil(t, err)
+
+	require.Equal(t, 2, len(response.Racks))
+
+// Racks is a map of a maps (A struct or array type). for eg: Rack1 is an map of blade 1 and blade2. blade 1 itself is a map.
+	
+	r,ok:= response.Racks["rack1"] // reading array value of  map Racks that was captured in response into a new variable r.
```
Comment (Jim3Things) [inline internal/config/InventoryReader_test.go:43]: For all cases in these two modules, please review and fix spacing around ':=', and add a space after the ','.  In this case, the line should be "r, ok := response.Racks["rack1"]"

File: internal/config/InventoryReader.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	pb "github.com/Jim3Things/CloudChamber/pkg/protos/inventory"
+)
+
+const(
+	DefaultDefinitionFile = "inventory.yaml"
+)
+
+//The local variables below were created to meet the requirements that would enable us to read the YAML file
+//zone is a local struct with a field as a map of Racks
+
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/InventoryReader.go:22]: Jim, in the 3rd sentence of the first comment, you ended with "Lay it out as..." and either the rest was truncated or it isn't displaying. Could you complete this thought?

File: internal/config/InventoryReader.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
+// Inventory reader parses the YAML file and returns Zone. into a pb external zone.
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/InventoryReader.go:1]: Jim, question. The "// Package <pkgname>" comment should apply to just one of the modules within the package presumably? I'm asking since elsewhere we seem to be missing a few of these and I was thinking of going hunting for and fixing some of these.

In this instance, assuming only a single "package" level comment should exist in the package, I'd suggest that the "package" level description probably belongs in settings.go but this module certainly needs at least a module description along the lines you suggested.

File: internal/config/InventoryReader.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+type tor struct {
+
+}
+
+type pdu struct {
+
+}
+
+func ReadInventoryDefinition(path string) (*pb.ExternalZone, error){
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/InventoryReader.go:53]: To expand on that, expect to add a comment for *all* public names (those beginning with a capital) otherwise the lint-er will complain. The "Problems" tab in VsCode will list the ones it has spotted with a "warning triangle" as an aid to locating them.

The comments should begin with "// <PublicName> just above the symbol declaration, or close by.

File: internal/config/InventoryReader.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+	viper.SetConfigName(DefaultDefinitionFile)
+	viper.AddConfigPath(path)
+	viper.SetConfigType(DefaultConfigType)
+
+	
+	tr := global.TraceProvider().Tracer("")
+
+	ctx, span :=tr.Start(
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/InventoryReader.go:62]: Jim, any suggested guidance here? I have a personal set but they really just boil down to "whatever is easier to read" which is not great help here.

File: internal/config/InventoryReader.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	// Fill in the global configuration object from the configuration file
+	xfr := &zone{}
+	if err := viper.UnmarshalExact(xfr); err != nil {
+		err = fmt.Errorf("unable to decode into struct, %v", err)
+		span.AddEvent(ctx, err.Error())
+		return nil, err
+	}
+	// function to translate xfr into an external zone
+	cfg := ToExternalZone(xfr)
+	span.AddEvent(ctx,
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/InventoryReader.go:93]: Also, the function "toExternalZone()" should be a private function, i.e. begin with a lowercase 't' as there is no need to access it from outside the package.

File: internal/config/InventoryReader.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/config/InventoryReader.go:1]: I honestly don't know, but I'm certainly happy with it in one place (doc.go, if we have that, settings.go in this case).

I'd suggest the module level comment should come after the package statement, just to differentiate it from the package level comment.

File: internal/config/InventoryReader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/config/InventoryReader.go:22]: Well, huh.  I did, and it got clipped.

What I'd written was:

section-comment
blank-line
type-comment
type-statement

File: internal/config/InventoryReader.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/config/InventoryReader.go:62]: For arguments on one line vs. stacked?  Yeah, my guidance is also kind of loose.  If the full line is <80 chars, I'd put it on one line.  Longer and I start looking to stack.

File: internal/config/InventoryReader.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/InventoryReader.go:1]: Works for me.

File: internal/config/InventoryReader.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/InventoryReader.go:62]: Good enough.

**Comment by manpreet85** [comment]
Function level comments added to the InventoryReader.go

**Review by MikeCJohnsonUk — APPROVED** [review]

**Review by Jim3Things — APPROVED** [review]
